presto-main/src/main/java/com/facebook/presto/operator/AggregationOperator.java

```diff
@@ -218,42 +272,122 @@ private Aggregator(AccumulatorFactory accumulatorFactory, Step step)
                 aggregation = accumulatorFactory.createIntermediateAccumulator();
             }
+
+            if (orderByChannels.isEmpty()) {
+                this.accumulator = aggregation;
+            }
+            else {
+                checkState(pagesIndexFactory.isPresent(), "No pagesIndexFactory to process ordering");
+                this.accumulator = new OrderingAccumulator(aggregation, orderBySourceTypes, orderByChannels, orderings, pagesIndexFactory.get());
+            }
+
             this.step = step;
         }
 
         public Type getType()
         {
             if (step.isOutputPartial()) {
-                return aggregation.getIntermediateType();
+                return accumulator.getIntermediateType();
             }
             else {
-                return aggregation.getFinalType();
+                return accumulator.getFinalType();
             }
         }
 
         public void processPage(Page page)
         {
             if (step.isInputRaw()) {
-                aggregation.addInput(page);
+                accumulator.addInput(page);
             }
             else {
-                aggregation.addIntermediate(page.getBlock(intermediateChannel));
+                accumulator.addIntermediate(page.getBlock(intermediateChannel));
             }
         }
 
         public void evaluate(BlockBuilder blockBuilder)
         {
             if (step.isOutputPartial()) {
-                aggregation.evaluateIntermediate(blockBuilder);
+                accumulator.evaluateIntermediate(blockBuilder);
             }
             else {
-                aggregation.evaluateFinal(blockBuilder);
+                accumulator.evaluateFinal(blockBuilder);
+            }
+        }
+
+        public long getEstimatedSize()
+        {
+            return accumulator.getEstimatedSize();
+        }
+    }
+
+    private static class OrderingAccumulator implements Accumulator
+    {
+        private final Accumulator accumulator;
+        private final List<Integer> orderByChannels;
+        private final List<SortOrder> orderings;
+        private final int[] orderByOutputChannels;
+        private final PagesIndex pagesIndex;
+
+        private OrderingAccumulator(
+                Accumulator accumulator,
+                List<Type> orderBySourceTypes,
+                List<Integer> orderByChannels,
+                List<SortOrder> orderings,
+                PagesIndex.Factory pagesIndexFactory)
+        {
+            this.accumulator = requireNonNull(accumulator, "accumulator is null");
+            this.orderByOutputChannels = new int[orderBySourceTypes.size()];
+            for (int i = 0; i < orderBySourceTypes.size(); i++) {
+                orderByOutputChannels[i] = i;
             }
+            this.orderByChannels = ImmutableList.copyOf(orderByChannels);
+            this.orderings = orderings;
+            this.pagesIndex = pagesIndexFactory.newPagesIndex(orderBySourceTypes, 10_000);
```

Comment:
Is `10_000` magical? I know it's already in some places in the code, but it looks more like copy&paste than a knowledge. Maybe this could be a default in pages index factory instead?